src/core/experimental/refcount.d

```diff
@@ -0,0 +1,548 @@
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope const ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a $(B const shared __RefCount) from a $(B const shared)
+    reference, `rhs`. This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope const shared ref typeof(this) rhs) const shared
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a $(B const __RefCount) from an `immutable` reference, `rhs`.
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) const
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Copy constructs a $(B const shared __RefCount) from an $(B immutable shared)
+    reference, `rhs`. This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable shared ref typeof(this) rhs) const shared
+    {
+        mixin(copyCtorIncRef);
+    }
+
+
+    /**
+    Copy construct an $(B immutable __RefCount) from an `immutable` reference, `rhs`.
+    This increases the reference count.
+    */
+    @nogc nothrow pure @safe scope
+    this(return scope immutable ref typeof(this) rhs) immutable
+    {
+        mixin(copyCtorIncRef);
+    }
+
+    /**
+    Assign a `__RefCount` object into this. This will decrement the old reference
+    count before assigning the new one. If the old reference was the last one,
+    this will trigger the deallocation of the old ref. This increases the
+    reference count of `rhs`.
+
+    Params:
+         rhs = the `__RefCount` object to be assigned.
+
+    Returns:
+         A reference to `this`.
+
+    Complexity:
+         $(BIGOH 1).
+    */
+    @nogc nothrow pure @safe scope
+    ref __RefCount opAssign(return scope ref typeof(this) rhs) return
+    {
+        if (rhs.isInitialized() && rc == rhs.rc)
+        {
+            return this;
+        }
+        if (rhs.isInitialized())
+        {
+            rhs.addRef();
+        }
+        if (isInitialized())
+        {
+            delRef();
+        }
+        () @trusted { rc = rhs.rc; }();
+        return this;
+    }
+
+    ///
+    @nogc nothrow pure @safe scope
+    ref shared(__RefCount) opAssign(return scope shared ref typeof(this) rhs) return shared
+    {
+        if (rhs.isInitialized() && rc == rhs.rc)
+        {
+            return this;
+        }
+        if (rhs.isInitialized())
+        {
+            rhs.addRef();
+        }
+        if (isInitialized())
+        {
+            delRef();
+        }
+        () @trusted { rc = rhs.rc; }();
+        return this;
+    }
+
+    /*
+    Increase the reference count. This asserts that `__RefCount` is initialized.
+
+    Returns:
+         This returns a `void*` so the compiler won't optimize away the call
+         to this `const pure` function.
```

Comment:
IIRC DMD never does this and LDC has a special attribute for this, though I could only find `@weak` :/

https://wiki.dlang.org/LDC-specific_language_changes

Reply:
why is this and rcOp marked pure in the first place? They very clearly aren't.

Reply:
Can we just get away with returning void then?

@wilzbach Are you sure LDC has an UDA for this? Can't find it here: https://github.com/ldc-developers/ldc/blob/eaa65fca18562dd71b4274a09763124cc2643934/gen/uda.cpp#L409

Reply:
@thewilsonator I'm not entirely sure, but I believe it should be weakly `pure` because it only modifies `this` and only calls other `pure` functions?

Reply:
> I believe it should be weakly `pure` because it only modifies `this` and only calls other `pure` functions?

If it modifies `this`, it's clearly not `const`.

Reply:
> If it modifies `this`, it's clearly not `const`.

You're right, it only uses `atomicOp` to increment/decrement `*rc` so it doesn't even modify `this`.

Reply:
> it only uses `atomicOp` to increment/decrement `*rc` so it doesn't even modify `this`

If you're reaching `*rc` via a const `this`, that's still invalid. Casting away const and mutating the data is not allowed. That includes const pointers to otherwise mutable data.

Reply:
I don't think there's really any way around that right now, until we get `__mutable`/`__metadata` or give the type special treatment (the `__` prefix is there in case things like this become an issue). Is this currently a problem on any compilers?

Reply:
Do we have reason to believe the language will definitely be changed to allow `__mutable`? Can we commit to something based on a future-truth that never happens? :/

Reply:
Like, i'm skeptical `__mutable` will ever appear; if `__mutable` is introduced, then any optimisation opportunity that `const` offers must immediately be explicitly denied by the spec... there's been tension on that issue for decades.